File: src/parser/lexer.ts
export type TokenType =
  | 'HEADING'
  | 'TEXT'
  | 'CODE_BLOCK'
  | 'LIST_ITEM'
  | 'EOF';

export interface Token {
  type: TokenType;
  value: string;
  line: number;
  column: number;
}

export class GrammarViolationError extends Error {
  constructor(message: string, public line: number, public column: number) {
    super(`Grammar Violation at ${line}:${column} - ${message}`);
    this.name = 'GrammarViolationError';
  }
}

export class Lexer {
  private input: string;
  private position: number = 0;
  private line: number = 1;
  private column: number = 1;

  constructor(input: string) {
    this.input = input;
  }

  private advance(): string {
    const char = this.input[this.position];
    this.position++;
    if (char === '\n') {
      this.line++;
      this.column = 1;
    } else {
      this.column++;
    }
    return char;
  }

  private peek(): string | null {
    if (this.position >= this.input.length) return null;
    return this.input[this.position];
  }

  public nextToken(): Token {
    if (this.position >= this.input.length) {
      return { type: 'EOF', value: '', line: this.line, column: this.column };
    }

    // Check for legacy syntax
    if (this.input.substring(this.position).trimStart().startsWith('domain "')) {
      throw new GrammarViolationError('Protocol Violation: Legacy .tela DSL syntax is deprecated.', this.line, this.column);
    }

    const startLine = this.line;
    const startColumn = this.column;
    const char = this.peek()!;

    if (char === '#') {
      let value = '';
      while (this.peek() === '#') {
        value += this.advance();
      }
      if (this.peek() === ' ') {
        this.advance(); // skip space
      }
      let headingText = '';
      while (this.peek() !== null && this.peek() !== '\n') {
        headingText += this.advance();
      }
      return { type: 'HEADING', value: headingText.trim(), line: startLine, column: startColumn };
    }

    if (this.input.substring(this.position).startsWith('```')) {
      this.advance(); this.advance(); this.advance(); // skip ```
      let value = '';
      while (this.peek() !== null && !this.input.substring(this.position).startsWith('```')) {
        value += this.advance();
      }
      if (this.peek() !== null) {
        this.advance(); this.advance(); this.advance(); // skip ```
      }
      return { type: 'CODE_BLOCK', value, line: startLine, column: startColumn };
    }

    if (char === '-' && this.input[this.position + 1] === ' ') {
      this.advance(); this.advance(); // skip "- "
      let value = '';
      while (this.peek() !== null && this.peek() !== '\n') {
        value += this.advance();
      }
      return { type: 'LIST_ITEM', value: value.trim(), line: startLine, column: startColumn };
    }

    // Text
    let value = '';
    while (this.peek() !== null && this.peek() !== '#' && !this.input.substring(this.position).startsWith('```') && !(this.peek() === '-' && this.input[this.position + 1] === ' ')) {
      value += this.advance();
    }
    
    // If value is just whitespace, skip it and get next token
    if (value.trim() === '') {
      return this.nextToken();
    }

    return { type: 'TEXT', value: value.trim(), line: startLine, column: startColumn };
  }

  public tokenize(): Token[] {
    const tokens: Token[] = [];
    while (true) {
      const token = this.nextToken();
      tokens.push(token);
      if (token.type === 'EOF') break;
    }
    return tokens;
  }
}

File: src/parser/lexer.rs
use std::iter::Peekable;
use std::str::Chars;

#[derive(Debug, PartialEq, Clone)]
pub enum Token {
    Keyword(String),
    String(String),
    Number(String),
    Ident(String),
    LBrace,
    RBrace,
    LBracket,
    RBracket,
    Colon,
    Comma,
    At,
    Eof,
}

pub struct Lexer<'a> {
    input: Peekable<Chars<'a>>,
    pub line: usize,
    pub column: usize,
}

impl<'a> Lexer<'a> {
    pub fn new(input: &'a str) -> Self {
        Self {
            input: input.chars().peekable(),
            line: 1,
            column: 1,
        }
    }

    fn advance(&mut self) -> Option<char> {
        let c = self.input.next();
        if let Some(ch) = c {
            if ch == '\n' {
                self.line += 1;
                self.column = 1;
            } else {
                self.column += 1;
            }
        }
        c
    }

    fn peek(&mut self) -> Option<&char> {
        self.input.peek()
    }

    fn skip_whitespace_and_comments(&mut self) {
        loop {
            match self.peek() {
                Some(&c) if c.is_whitespace() => {
                    self.advance();
                }
                Some(&'/') => {
                    // Peek ahead without advancing if possible, but Peekable only peeks 1.
                    // We'll just clone the iterator to peek ahead.
                    let mut lookahead = self.input.clone();
                    lookahead.next(); // skip '/'
                    if let Some(&'/') = lookahead.peek() {
                        self.advance(); // consume '/'
                        self.advance(); // consume '/'
                        while let Some(&c) = self.peek() {
                            if c == '\n' {
                                break;
                            }
                            self.advance();
                        }
                    } else {
                        break;
                    }
                }
                Some(&'%') => {
                    let mut lookahead = self.input.clone();
                    lookahead.next(); // skip '%'
                    if let Some(&'%') = lookahead.peek() {
                        self.advance(); // consume '%'
                        self.advance(); // consume '%'
                        while let Some(&c) = self.peek() {
                            if c == '\n' {
                                break;
                            }
                            self.advance();
                        }
                    } else {
                        break;
                    }
                }
                Some(&'#') => {
                    self.advance(); // consume '#'
                    while let Some(&c) = self.peek() {
                        if c == '\n' {
                            break;
                        }
                        self.advance();
                    }
                }
                _ => break,
            }
        }
    }

    pub fn next_token(&mut self) -> Result<Token, String> {
        self.skip_whitespace_and_comments();

        let c = match self.peek() {
            Some(&ch) => ch,
            None => return Ok(Token::Eof),
        };

        match c {
            '{' => {
                self.advance();
                Ok(Token::LBrace)
            }
            '}' => {
                self.advance();
                Ok(Token::RBrace)
            }
            '[' => {
                self.advance();
                Ok(Token::LBracket)
            }
            ']' => {
                self.advance();
                Ok(Token::RBracket)
            }
            ':' => {
                self.advance();
                Ok(Token::Colon)
            }
            ',' => {
                self.advance();
                Ok(Token::Comma)
            }
            '@' => {
                self.advance();
                Ok(Token::At)
            }
            '"' => self.read_string(),
            c if c.is_ascii_digit() || c == '-' || c == '.' => self.read_number(),
            c if c.is_alphabetic() || c == '_' => self.read_ident_or_keyword(),
            _ => {
                let err = format!("Unexpected character '{}' at line {}, col {}", c, self.line, self.column);
                self.advance();
                Err(err)
            }
        }
    }

    pub fn tokenize(&mut self) -> Result<Vec<Token>, String> {
        let mut tokens = Vec::new();
        loop {
            let token = self.next_token()?;
            let is_eof = token == Token::Eof;
            tokens.push(token);
            if is_eof {
                break;
            }
        }
        Ok(tokens)
    }

    fn read_string(&mut self) -> Result<Token, String> {
        self.advance(); // skip opening quote
        let mut is_multiline = false;

        // Check for multiline string """
        if let Some(&'"') = self.peek() {
            self.advance();
            if let Some(&'"') = self.peek() {
                self.advance();
                is_multiline = true;
            } else {
                // Empty string ""
                return Ok(Token::String(String::new()));
            }
        }

        let mut result = String::new();
        if is_multiline {
            let mut quotes_count = 0;
            loop {
                match self.advance() {
                    Some('"') => {
                        quotes_count += 1;
                        if quotes_count == 3 {
                            break;
                        }
                    }
                    Some(c) => {
                        for _ in 0..quotes_count {
                            result.push('"');
                        }
                        quotes_count = 0;
                        result.push(c);
                    }
                    None => return Err(format!("Unterminated multiline string starting at line {}", self.line)),
                }
            }
        } else {
            loop {
                match self.advance() {
                    Some('"') => break,
                    Some('\\') => {
                        match self.advance() {
                            Some('n') => result.push('\n'),
                            Some('r') => result.push('\r'),
                            Some('t') => result.push('\t'),
                            Some('\\') => result.push('\\'),
                            Some('"') => result.push('"'),
                            Some(c) => {
                                result.push('\\');
                                result.push(c);
                            }
                            None => return Err("Unterminated string escape".to_string()),
                        }
                    }
                    Some(c) => result.push(c),
                    None => return Err(format!("Unterminated string starting at line {}", self.line)),
                }
            }
        }

        Ok(Token::String(result))
    }

    fn read_number(&mut self) -> Result<Token, String> {
        let mut result = String::new();
        while let Some(&c) = self.peek() {
            if c.is_ascii_digit() || c == '.' || c == '-' {
                result.push(c);
                self.advance();
            } else {
                break;
            }
        }
        Ok(Token::Number(result))
    }

    fn read_ident_or_keyword(&mut self) -> Result<Token, String> {
        let mut result = String::new();
        while let Some(&c) = self.peek() {
            if c.is_alphanumeric() || c == '_' {
                result.push(c);
                self.advance();
            } else {
                break;
            }
        }

        match result.as_str() {
            "domain" | "meta" | "feature" | "deterministic" | "sustainability" => Ok(Token::Keyword(result)),
            _ => Ok(Token::Ident(result)),
        }
    }
}
